dbt/adapters/redshift/impl.py

```diff
@@ -17,6 +19,76 @@ def type(cls):
     def date_function(cls):
         return 'getdate()'
 
+    @classmethod
+    def get_redshift_credentials(cls, config):
+        result = config.copy()
+
+        method = result.get('method')
+
+        if method == 'database':
+            return (result)
+
+        elif method == 'iam':
+            cluster_id = result.get('cluster_id')
+            if not cluster_id:
+                error = '`cluster_id` must be set in profile if IAM authentication method selected'
+                raise dbt.exceptions.FailedToConnectException(error)
+
+            client = boto3.client('redshift')
+
+            # replace username and password with temporary redshift credentials
+            try:
+                cluster_creds = client.get_cluster_credentials(DbUser=result.get('user'),
+                                                               DbName=result.get('dbname'),
+                                                               ClusterIdentifier=result.get('cluster_id'),
+                                                               AutoCreate=False)
+                result['user_tmp'] = cluster_creds.get('DbUser')
+                result['pass_tmp'] = cluster_creds.get('DbPassword')
+            except client.exceptions.ClientError as e:
+                error = ('Unable to get temporary Redshift cluster credentials: "{}"'.format(str(e)))
+                raise dbt.exceptions.FailedToConnectException(error)
+
+            return result
+
+        else:
+            error = ('Invalid `method` in profile: "{}"'.format(method))
+            raise dbt.exceptions.FailedToConnectException(error)
+
+    @classmethod
+    def open_connection(cls, connection):
+        if connection.get('state') == 'open':
+            logger.debug('Connection is already open, skipping open.')
+            return connection
+
+        result = connection.copy()
+
+        try:
+            credentials = cls.get_redshift_credentials(connection.get('credentials', {}))
```

Comment:
the suggestion above about `get_redshift_credentials` should clean up the if/else logic here I believe!

Reply:
Done.